src/Microsoft.AspNetCore.Hosting/Internal/HostingApplicationDiagnostics.cs

```diff
@@ -0,0 +1,269 @@
+            {
+                if (startTimestamp == 0)
+                {
+                    startTimestamp = Stopwatch.GetTimestamp();
+                }
+
+                // Non-inline
+                LogRequestStarting(httpContext);
+            }
+
+            context.StartTimestamp = startTimestamp;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void RequestEnd(HttpContext httpContext, Exception exception, HostingApplication.Context context)
+        {
+            // Local cache items resolved multiple items, in order of use so they are primed in cpu pipeline when used
+            var startTimestamp = context.StartTimestamp;
+            long currentTimestamp;
+
+            if (_diagnosticListener.IsEnabled())
+            {
+                currentTimestamp = Stopwatch.GetTimestamp();
+
+                if (exception == null)
+                {
+                    // No exception was thrown, request was sucessful
+                    if (_diagnosticListener.IsEnabled(DeprecatedDiagnosticsEndRequestKey))
+                    {
+                        // Diagnostics is enabled for EndRequest, but it may not be for BeginRequest
+                        // so call GetTimestamp if currentTimestamp is zero (from above)
+                        RecordEndRequestDiagnostics(httpContext, currentTimestamp);
+                    }
+                }
+                else
+                {
+                    // Exception was thrown from request
+                    if (_diagnosticListener.IsEnabled(DeprecatedDiagnosticsUnhandledExceptionKey))
+                    {
+                        // Diagnostics is enabled for UnhandledException, but it may not be for BeginRequest
+                        // so call GetTimestamp if currentTimestamp is zero (from above)
+                        RecordUnhandledExceptionDiagnostics(httpContext, currentTimestamp, exception);
+                    }
+
+                }
+
+                var activity = context.Activity;
+                // Always stop activity if it was started
+                if (activity != null)
+                {
+                    StopActivity(httpContext, activity);
+                }
+            }
+
+            if (context.EventLogEnabled && exception != null)
+            {
+                // Non-inline
+                HostingEventSource.Log.UnhandledException();
+            }
+
+            // If startTimestamp was 0, then Information logging wasn't enabled at for this request (and calcuated time will be wildly wrong)
+            // Is used as proxy to reduce calls to virtual: _logger.IsEnabled(LogLevel.Information)
+            if (startTimestamp != 0)
+            {
+                if (currentTimestamp == 0)
+                {
+                    currentTimestamp = Stopwatch.GetTimestamp();
+                }
+                // Non-inline
+                LogRequestFinished(httpContext, startTimestamp, currentTimestamp);
```

Comment:
could you call `LogRequestFinished` before stopping Activity. This way some custom ILogger implementation may just use Activity.Current.Id and other properties along with the log record